# 7. Análisis utilizando Tipos de Datos Complejos

In [2]:
# @title Postgress preloading (install server)
#The output of the installation is not displayed when %%capture is used at the start of the cell
%%capture
# Install postgresql server
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start
# Setup a password `postgres` for username `postgres`
!sudo -u postgres psql -U postgres -c "ALTER USER postgres PASSWORD 'postgres';"
# Setup a database with name `sampledb` to be used
# !sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS sampledb;'
# !sudo -u postgres psql -U postgres -c 'CREATE DATABASE sampledb;'

!curl https://raw.githubusercontent.com/limspiga/data-modeling/main/db/data.dump  -O
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE sqlda;'

# !sudo -u postgres psql -U postgres -d 'sqlda' -f 'data.dump'

# https://thivyapriyaa.medium.com/setting-up-postgresql-on-google-colab-4d02166939fc


In [3]:
# @title Postgress preloading
# import
!sudo -u postgres psql -d sqlda < data.dump
%env DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda
#To load the sql extention to start using %%sql
%load_ext sql

SET
SET
SET
SET
SET
 set_config 
------------
 
(1 row)

SET
SET
SET
SET
CREATE EXTENSION
COMMENT
CREATE EXTENSION
COMMENT
CREATE TEXT SEARCH DICTIONARY
ALTER TEXT SEARCH DICTIONARY
SET
SET
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
COPY 44533
COPY 0
COPY 50000
COPY 32
COPY 50000
COPY 20
COPY 418158
COPY 12
COPY 15412
COPY 37711
COPY 300
COPY 20
ALTER TABLE
ALTER TABLE
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
REFRESH MATERIALIZED VIEW
REFRESH MATERIALIZED VIEW
env: DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda


# Introducción
En este curso, has aprendido mucho sobre el poder de procesamiento de SQL en números y cadenas de caracteres. La mayoría de las tareas de análisis de datos consisten, de hecho, en analizar números y cadenas de caracteres. Sin embargo, en el mundo real, los datos a menudo se encuentran en otros formatos, como
- palabras,
- ubicaciones,
- fechas y, a veces,
- estructuras de datos complejas.

Estos datos, aunque se presenten como números y cadenas de caracteres, tienen su propio dominio de operación y cálculo en lugar de la simple aritmética.
- Por ejemplo, sumar un día a enero 31, 2022, dará como resultado febrero 1, 2022, no enero 32, 2022.

En este capítulo, examinarás estos tipos de datos y verás cómo puedes utilizar estos datos en tu análisis:

- Fecha y hora
- Geoespacial
- JSON
- ARRAY
- Texto

Al final del capítulo, habrás ampliado tus capacidades de análisis para que puedas aprovechar prácticamente cualquier tipo de dato disponible para ti.

# Tipos de Datos de Fecha y Hora para Análisis
Puede que estés familiarizado con las fechas y las horas, pero ¿sabes cómo se representan estas medidas cuantitativas? Se representan usando números, pero no un solo número. En su lugar, se miden con un conjunto de números, con un número para cada año, mes, día, hora, minuto, segundo y milisegundo.

Esta es una representación compleja, que comprende varios componentes diferentes. Por ejemplo, conocer el minuto actual sin saber la hora actual no sirve de nada. Además, hay formas complejas de interactuar con las fechas y las horas; por ejemplo, se pueden restar diferentes puntos en el tiempo entre sí. La hora actual puede representarse de manera diferente dependiendo de dónde te encuentres en el mundo.

Como resultado de estas intrincadas, necesitas tener especial cuidado al trabajar con este tipo de datos. De hecho, PostgreSQL, como la mayoría de las bases de datos, ofrece tipos de datos especiales que pueden representar estos tipos de valores. Comenzarás examinando el tipo DATE.

## El tipo de dato DATE
Las fechas pueden ser fácilmente representadas usando cadenas o números
- (por ejemplo, tanto el 1 de enero de 2022 como el 01/01/2022 representan claramente una fecha específica),

pero las fechas son una forma especial de datos ya que representan un valor cuantitativo que no siempre sigue la simple secuencia numérica.

- Agregar 1 semana a la fecha actual significa agregar 7 días, por ejemplo.

Una fecha dada tiene diferentes propiedades que podrías querer usar en tu análisis, por ejemplo, el año o el día de la semana que la fecha representa. Trabajar con fechas también es necesario para el análisis de series temporales, que es uno de los tipos de análisis más comunes que surgen. El estándar SQL incluye el tipo de dato DATE.

La preocupación más común sobre el tipo de dato `DATE` es el formato de visualización.
Diferentes regiones utilizan diferentes formatos para representar la misma fecha.

- Por ejemplo, la fecha 14 de enero de 2022 se escribe como 01/14/2022 en algunos países pero 14/01/2022 en otros. Puedes configurar tu base de datos para mostrar las fechas en el formato que te sea más familiar. PostgreSQL utiliza el parámetro `DateStyle` para configurar estas opciones. Para ver tu configuración actual, puedes usar el siguiente comando:

In [4]:
 %%sql
 SHOW DateStyle;

1 rows affected.


[('ISO, MDY',)]

La primera salida especifica el formato de salida de la Organización Internacional de Normalización (ISO), que muestra la fecha como Mes-Día-Año, y el segundo parámetro de salida especifica el orden para la entrada o salida. En este caso, dado que tanto el servidor como el cliente están utilizando una configuración regional de Estados Unidos, se utiliza el estilo de visualización Mes/Día/Año. Si tu servidor o cliente PostgreSQL está instalado en un sistema operativo con una configuración regional del sistema diferente a la mencionada aquí, el resultado del comando anterior puede ser diferente. Por ejemplo, si quisieras configurarlo al formato europeo de Día, Mes, Año, configurarías DateStyle a 'GERMAN, DMY'. Puedes configurar la salida para tu base de datos utilizando el siguiente comando:

In [5]:
%%sql
-- SET DateStyle='GERMAN, DMY';
SET DateStyle = 'ISO, MDY';

 * postgresql://postgres:***@localhost:5432/sqlda
Done.


[]

Para este capítulo, utilizarás el formato de visualización ISO (Año-Mes-Día) y el formato de entrada Mes/Día/Año. Puedes configurar este formato utilizando el comando anterior.
Ahora, comienza probando el formato DATE:

In [6]:
%sql SELECT '02-01-23'::DATE;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.date(2023, 2, 1),)]

In [7]:
%sql SELECT TO_DATE('1/1/2022', 'D/M/YYYY');

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.date(2022, 1, 1),)]

El tipo de dato DATE es útil. La extensión natural de este es el tipo de dato que representa la hora, como las 10 a.m. o las 2 p.m., en un día. Lo interesante es que cuando las personas hablan sobre la hora, usualmente se refieren a la combinación de un día y una hora. Simplemente referirse a una hora no es suficiente para determinar el momento exacto en que algo sucede. Por ejemplo, "la clase comienza a las 6 p.m." muy probablemente implica que la clase comienza a las 6 p.m. todos los lunes de este semestre. Para evitar cualquier confusión, el estándar SQL ofrece el tipo de dato TIMESTAMP, que representa una fecha y una hora, hasta un microsegundo, por ejemplo, 2022-06-05 13:47:44.472096.

Puedes ver la marca de tiempo actual usando la función NOW(), y puedes especificar tu zona horaria usando AT TIME ZONE '<zona horaria>'. Aquí hay un ejemplo de la función NOW() con la zona horaria del Este especificada:




In [9]:
%sql SELECT NOW() AT TIME ZONE 'EST';

# https://www.postgresql.org/docs/7.2/timezones.html

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.datetime(2023, 11, 10, 17, 14, 37, 636153),)]

También puedes usar el tipo de dato TIMESTAMP sin especificar la zona horaria. Puedes obtener la zona horaria actual con la función NOW():

In [10]:
%sql SELECT NOW();

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.datetime(2023, 11, 10, 22, 14, 55, 229967, tzinfo=datetime.timezone.utc),)]

# Transformando Tipos de Datos de Fecha
A menudo, querrás descomponer tus fechas en sus partes componentes. Por ejemplo, aunque las ventas diarias son importantes, también podrías estar interesado en el resumen de cada año y mes, para que puedas revisar la tendencia mensual de tus ventas. Puedes ver a partir de esta tendencia cuál mes es el que tiene las mejores ventas, para que puedas preparar tu inventario con anticipación. Para hacer esto, puedes usar EXTRACT(componente FROM fecha). Aquí hay un ejemplo:

In [11]:
%%sql
SELECT
  current_date,
  EXTRACT(year FROM current_date) AS year,
  EXTRACT(month FROM current_date) AS month,
  EXTRACT(day FROM current_date) AS day;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.date(2023, 11, 10), Decimal('2023'), Decimal('11'), Decimal('10'))]

In [12]:
%%sql
SELECT
  current_date,
  EXTRACT(y FROM current_date) AS year,
  EXTRACT(mon FROM current_date) AS month,
  EXTRACT(d FROM current_date) AS day;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.date(2023, 11, 10), Decimal('2023'), Decimal('11'), Decimal('10'))]

Además del año, mes y día, en ocasiones desearás obtener componentes adicionales, como el día de la semana, la semana del año o el trimestre. Puedes extraer estas partes de la fecha de la siguiente manera:

In [13]:
%%sql
SELECT
  current_date,
  EXTRACT(dow FROM current_date) AS day_of_week,
  EXTRACT(week FROM current_date) AS week_of_year,
  EXTRACT(quarter FROM current_date) AS quarter;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.date(2023, 11, 10), Decimal('5'), Decimal('45'), Decimal('4'))]

Además de extraer partes de una fecha de una fecha, es posible que desees simplemente truncar tu fecha o timestamp. Por ejemplo, supongamos que deseas ver el resumen de ventas del año y mes, pero solo tienes la fecha en la tabla de ventas. Para agregar las ventas por año/mes, necesitas eliminar el día y el timestamp de la fecha y obtener la salida de año+mes. Esto se puede hacer usando muchas funciones, como DATE_TRUNC(), DATE_PART(), o EXTRACT(), cada una con una sintaxis y propósito ligeramente diferente. En el siguiente ejemplo, utilizarás la función TO_CHAR(), que extrae las partes designadas de una fecha y las organiza en una cadena, porque ofrece máxima flexibilidad sobre qué información puedes obtener y cómo deseas presentarla:

In [14]:
%%sql
SELECT NOW(), TO_CHAR(NOW(), 'yyyymm') AS yearmonth;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.datetime(2023, 11, 10, 22, 15, 59, 981881, tzinfo=datetime.timezone.utc), '202311')]

Las funciones de extracción de partes de fecha como TO_CHAR() y EXTRACT() son particularmente útiles para las instrucciones GROUP BY. Por ejemplo, puedes usarlo para agrupar las ventas por año y mes y obtener las ventas mensuales para todo el año.

In [15]:
%%sql
SELECT
  TO_CHAR(sales_transaction_date, 'yyyymm') AS yearmonth,
  SUM(sales_amount) AS total_quarterly_sales
FROM sales
GROUP BY 1
ORDER BY
  1 DESC LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[('202201', 5706307.722000058),
 ('202112', 5455222.157000077),
 ('202111', 4280693.541000046),
 ('202110', 3318119.411000051),
 ('202109', 3509884.9920000336),
 ('202108', 3937509.896000073),
 ('202107', 5485778.13700009),
 ('202106', 5065843.292000066),
 ('202105', 6948262.514000091),
 ('202104', 6913306.458000054)]

## Intervalos
Además de representar fechas, también puedes representar intervalos de tiempo fijos utilizando el tipo de dato INTERVAL. Este tipo de dato es útil si deseas analizar cuánto tiempo tarda algo. Por ejemplo, cuando los clientes reciben un correo electrónico promocional, pueden no abrirlo inmediatamente. El intervalo entre la fecha en que se recibe el correo electrónico y la fecha en que se abre puede indicar cuán atractivo es el correo electrónico para los clientes. Si quieres saber cuánto tiempo tarda un cliente en abrir un correo electrónico después de recibirlo, necesitas calcular el intervalo entre esas dos fechas.
Aquí hay un ejemplo:

In [23]:
%sql SELECT INTERVAL '5 days';

%sql SELECT now()  - INTERVAL '7 days';

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.
 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.datetime(2023, 11, 3, 22, 19, 32, 934889, tzinfo=datetime.timezone.utc),)]

Los intervalos son útiles para medir la diferencia entre dos marcas de tiempo al restar estas dos marcas de tiempo. También se pueden utilizar para sumar a o restar de una marca de tiempo para obtener una nueva marca de tiempo. Por ejemplo, si desea medir la longitud de febrero, puede calcular el intervalo entre el primer día de febrero y el primer día de marzo:

In [ ]:
%%sql
SELECT TIMESTAMP '2022-03-01 00:00:00' - TIMESTAMP '2022-02-01
00:00:00' AS days_in_feb;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.timedelta(days=28),)]

Alternativamente, los intervalos pueden usarse para sumar la cantidad de días a una marca de tiempo para obtener una nueva marca de tiempo, como "¿cuál es la fecha dentro de 7 días a partir de ahora?":

In [ ]:
%sql SELECT TIMESTAMP '2022-06-05 00:00:00' + INTERVAL '7 days' AS new_date;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.datetime(2022, 6, 12, 0, 0),)]

Aunque los intervalos ofrecen un método preciso para realizar aritmética con marcas de tiempo, el formato DATE puede ser usado con enteros para lograr un resultado similar. En el siguiente ejemplo, simplemente sumas 7 (un entero) a la fecha para calcular la nueva fecha:

In [ ]:
%sql SELECT DATE '2022-06-05' + 7 AS new_date;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(datetime.date(2022, 6, 12),)]

De manera similar, puedes restar dos fechas y obtener un resultado en enteros:

In [ ]:
%sql SELECT DATE '2022-03-01' - DATE '2022-02-01' AS days_in_feb;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(28,)]

Aunque el tipo de datos DATE ofrece facilidad de uso, el tipo de datos timestamp con TIME ZONE ofrece precisión. Si necesitas que tu campo de fecha/hora sea precisamente el mismo que el momento en el que ocurrió la acción, deberías usar el timestamp con la zona horaria. Si no, puedes usar el campo DATE.

# Ejercicio 7.01: Análisis con Datos de Series Temporales
ZoomZoom ha intensificado sus esfuerzos para reclutar más clientes durante el año 2021, esperando que pueda vender más vehículos a medida que crezca el número de nuevos clientes. En este ejercicio, realizarás un análisis básico utilizando datos de series temporales para obtener información sobre si las ventas se vieron afectadas por el número de nuevos clientes. Aunque tiene sentido hacer una comparación día a día, las ventas/reclutamientos diarios pueden fluctuar significativamente. Generalmente se recomienda comenzar desde un período de tiempo más largo, como ventas/reclutamientos mensuales, y desglosar los números una vez que encuentres algún patrón.
Realiza los siguientes pasos para completar el ejercicio:
1. Primero, observa el número de ventas mensuales. Puedes usar la siguiente consulta agregada con el método TO_CHAR:

In [24]:
%%sql
SELECT
  TO_CHAR(sales_transaction_date, 'yyyymm') AS month_date,
  COUNT(1) AS number_of_sales
FROM sales
WHERE
  EXTRACT(year FROM sales_transaction_date) = 2021
GROUP BY 1
ORDER BY 1;

 * postgresql://postgres:***@localhost:5432/sqlda
12 rows affected.


[('202101', 989),
 ('202102', 929),
 ('202103', 1111),
 ('202104', 994),
 ('202105', 887),
 ('202106', 765),
 ('202107', 788),
 ('202108', 593),
 ('202109', 585),
 ('202110', 634),
 ('202111', 736),
 ('202112', 893)]

2. Ejecute otra consulta para obtener el número de nuevos clientes que se unen cada mes:

In [25]:
%%sql
SELECT
  TO_CHAR(date_added, 'yyyymm') AS month_date,  -- solo cuando tiene valor en campo
  COUNT(1) AS number_of_new_customers
FROM
  customers
WHERE
  EXTRACT(year FROM date_added) = 2021
GROUP BY 1
ORDER BY 1;

 * postgresql://postgres:***@localhost:5432/sqlda
12 rows affected.


[('202101', 466),
 ('202102', 439),
 ('202103', 455),
 ('202104', 433),
 ('202105', 463),
 ('202106', 439),
 ('202107', 482),
 ('202108', 467),
 ('202109', 421),
 ('202110', 447),
 ('202111', 474),
 ('202112', 507)]

Probablemente puedas ver que el flujo de nuevos clientes potenciales es bastante estable y oscila alrededor de 400-500 nuevas inscripciones de clientes cada mes, mientras que el número de ventas (como se consulta en el paso 1) varía considerablemente. Entonces, parece que el esfuerzo de inscribir a nuevos clientes puede no estar directamente relacionado con la cantidad de ventas.

En este ejercicio, utilizaste una función de PostgreSQL para extraer las partes de año y mes de una fecha y utilizaste la información extraída para agregar datos de ventas y datos de reclutamiento de clientes para formar una serie temporal para comparación. En la siguiente sección, aprenderás sobre otro tipo de datos que tiene su propio dominio de operaciones, los datos geoespaciales.

# Realizando Análisis Geoespacial en PostgreSQL
Además de observar datos de series temporales para comprender mejor las tendencias, también puedes utilizar información geoespacial (como ciudad, país o latitud y longitud) para entender mejor a tus clientes. Por ejemplo, los gobiernos utilizan análisis geoespacial para comprender mejor las diferencias económicas regionales, mientras que una plataforma de viajes compartidos podría utilizar datos geoespaciales para encontrar el conductor más cercano para un cliente.

Puedes representar una ubicación geoespacial utilizando coordenadas de latitud y longitud, y esto será el bloque fundamental para comenzar el análisis geoespacial.

## Latitud y Longitud
Las ubicaciones a menudo se piensan en términos de la dirección: la ciudad, estado, país o código postal asignado a la ubicación que te interesa. Esto es generalmente desde una perspectiva analítica. Por ejemplo, puedes observar el volumen de ventas en la tabla de ventas de ZoomZoom por ciudad y obtener resultados significativos sobre qué ciudades están teniendo un buen desempeño.

A menudo, necesitas entender las relaciones geoespaciales numéricamente para comprender las distancias entre dos puntos o relaciones que varían según donde te encuentres en un mapa. Después de todo, si vives en la frontera entre dos ciudades, es improbable que tu comportamiento de gasto cambie repentinamente si cruzas a la otra ciudad.

 La latitud y la longitud te permiten observar la ubicación en un contexto continuo. Esto te permite analizar las relaciones numéricas entre la ubicación y otros factores (por ejemplo, las ventas). La latitud y la longitud también te permiten observar las distancias entre dos ubicaciones.

- La latitud te indica cuán lejos al norte o al sur está un punto. Un punto en +90° de latitud está en el Polo Norte, mientras que un punto en 0° de latitud está en el ecuador, y un punto en -90° está en el Polo Sur. En un mapa, las líneas de latitud constante corren de este a oeste.


- La longitud te indica cuán lejos al este o al oeste está un punto. En un mapa, las líneas de longitud constante corren de norte a sur. Greenwich, Inglaterra, es el punto de 0° de longitud. Los puntos pueden definirse usando la longitud como oeste (-) o este (+) de este punto y los valores varían desde -180° oeste hasta +180° este. Estos valores son equivalentes porque ambos apuntan a la línea vertical que atraviesa el Océano Pacífico, que está a medio camino alrededor del mundo desde Greenwich, Inglaterra.



<figure>
<center><img src='https://mixdyr.files.wordpress.com/2020/07/coordenadas_geograficas_rfsm.jpg' width="400" />
<figcaption>
Figure 7.1: Corordenadas geográficas</figcaption></center>
</figure>

## Representando Latitud y Longitud en PostgreSQL

En PostgreSQL, puedes representar latitud y longitud usando dos números de punto flotante. De hecho, así es como se representan la latitud y la longitud en la tabla de clientes de ZoomZoom:

In [26]:
%%sql
SELECT
  latitude,
  longitude
FROM
  customers
WHERE
  latitude IS NOT NULL
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(31.6948, -106.3),
 (30.6589, -88.178),
 (40.0086, -76.5972),
 (38.4142, -81.7582),
 (47.1441, -122.4434),
 (36.3551, -119.301),
 (38.8318, -77.2888),
 (39.5645, -75.597),
 (38.5814, -90.2625),
 (30.6143, -87.2758)]

Aquí, puedes ver que todas las latitudes son positivas porque Estados Unidos está al norte del ecuador. Todas las longitudes son negativas porque Estados Unidos está al oeste de Greenwich, Inglaterra. También puedes notar que algunos clientes no tienen valores de latitud y longitud llenados, porque su ubicación es desconocida.

Aunque estos valores pueden proporcionarte la ubicación exacta de un cliente, no puedes hacer mucho con esa información, porque los cálculos de distancia requieren **trigonometría** y hacen suposiciones simplificadas de que la Tierra es perfectamente redonda.

Afortunadamente, PostgreSQL tiene las herramientas para resolver este problema. Puedes calcular distancias en PostgreSQL usando dos paquetes, earthdistance y cube. Puedes instalar estos dos paquetes ejecutando los siguientes dos comandos en pgAdmin:

In [27]:
%%sql
CREATE EXTENSION cube;
CREATE EXTENSION earthdistance;

 * postgresql://postgres:***@localhost:5432/sqlda
(psycopg2.errors.DuplicateObject) extension "cube" already exists

[SQL: CREATE EXTENSION cube;]
(Background on this error at: https://sqlalche.me/e/20/f405)


Estas dos extensiones solo necesitan ser instaladas una vez ejecutando los dos comandos precedentes. El módulo earthdistance depende del módulo cube, así que debes instalar el módulo cube primero. Una vez que instales el módulo earthdistance, puedes definir un tipo de datos POINT:

In [28]:
%%sql
-- Figure 7.5: Customer latitude and longitude represented as points in PostgreSQL
SELECT
  POINT(longitude, latitude)
FROM
  customers
WHERE
  longitude IS NOT NULL
LIMIT 10;


 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[('(-106.3,31.6948)',),
 ('(-88.178,30.6589)',),
 ('(-76.5972,40.0086)',),
 ('(-81.7582,38.4142)',),
 ('(-122.4434,47.1441)',),
 ('(-119.301,36.3551)',),
 ('(-77.2888,38.8318)',),
 ('(-75.597,39.5645)',),
 ('(-90.2625,38.5814)',),
 ('(-87.2758,30.6143)',)]

El módulo earthdistance también te permite calcular la distancia entre puntos en millas:

In [29]:
%%sql
SELECT
 point(-90, 38) <@> point(-91, 37) AS distance_in_miles;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(88.19493383797524,)]

In [30]:
%%sql
SELECT
 format('%s', ((point(-90, 38) <@> point(-91, 37)) * 1609.34)::numeric(10,2)) AS distance_in_meters;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[('141935.63',)]

En este ejemplo, definiste dos puntos, (38° N, 90° O) y (37° N, 91° O), y pudiste calcular la distancia entre estos puntos usando el operador <@>. Este operador calcula la distancia en millas (en este caso, estos dos puntos están a 88.2 millas de distancia).

En el siguiente ejercicio, verás cómo puedes utilizar estos cálculos de distancia en un contexto empresarial práctico.

# Ejercicio 7.02: Análisis Geoespacial
En este ejercicio, identificarás el concesionario más cercano para cada cliente. Los comercializadores de ZoomZoom están tratando de aumentar la participación del cliente ayudando a los clientes a encontrar su concesionario más cercano. Al equipo de producto también le interesa saber cuál es la distancia típica entre cada cliente y su concesionario más cercano.

1. Crea una tabla con los puntos de longitud y latitud para cada cliente:

In [31]:
%%sql
CREATE TEMP TABLE customer_points AS (
  SELECT
    customer_id,
point(longitude, latitude) AS lng_lat_point
  FROM
customers
  WHERE
longitude IS NOT NULL
  AND
    latitude IS NOT NULL
);

 * postgresql://postgres:***@localhost:5432/sqlda
44533 rows affected.


[]

2. Crea una tabla similar para cada concesionario:


In [32]:
%%sql
CREATE TEMP TABLE dealership_points AS (
  SELECT
    dealership_id,
    point(longitude, latitude) AS lng_lat_point
    FROM
    dealerships
);

 * postgresql://postgres:***@localhost:5432/sqlda
20 rows affected.


[]

3. Realiza un cruce entre estas tablas para calcular la distancia desde cada cliente hasta cada concesionario (en millas):


In [41]:
# %sql DROP TABLE  customer_dealership_distance;

%%sql
CREATE TEMP TABLE  customer_dealership_distance AS (
  SELECT
    customer_id,
    dealership_id,
    c.lng_lat_point <@> d.lng_lat_point AS distance
  FROM
    customer_points c
  CROSS JOIN -- IMPORTANTE TODOS LOS CLIENTES CON TODOS LO CONCESIONARIOS
    dealership_points d
);

 * postgresql://postgres:***@localhost:5432/sqlda
890660 rows affected.


[]

4. Finalmente, para cada ID de cliente, selecciona el concesionario con la menor distancia. Hasta ahora, has obtenido la ubicación para los clientes y concesionarios y las distancias desde cada cliente hasta cada concesionario. La siguiente tarea es encontrar la combinación cliente-concesionario que tiene la menor distancia para el cliente utilizando una cláusula DISTINCT ON. Como se discutió en el Capítulo 6, Importar y Exportar Datos, la cláusula DISTINCT ON garantiza solo el primer registro para cada valor único de la columna entre paréntesis. En este caso, obtendrás un registro para cada valor de customer_id, y dado que esto está ordenado por distancia a los concesionarios, obtendrás el concesionario que tiene la menor distancia:

In [44]:
# %sql DROP TABLE  closest_dealerships;
%%sql
CREATE TEMP TABLE closest_dealerships AS (
  SELECT DISTINCT ON (customer_id) -- SOLO TOMA EL PRIMER REGISTRO POR CUSTOMER
    customer_id,
    dealership_id,
    distance
  FROM customer_dealership_distance
  ORDER BY customer_id, distance
);

 * postgresql://postgres:***@localhost:5432/sqlda
44533 rows affected.


[]

In [43]:
%%sql
SELECT DISTINCT ON (customer_id)
  customer_id,
  dealership_id,
  distance
FROM customer_dealership_distance
ORDER BY customer_id, distance LIMIT 12

 * postgresql://postgres:***@localhost:5432/sqlda
12 rows affected.


[(2, 18, 231.80698588435425),
 (3, 11, 274.9885104949864),
 (4, 11, 208.93337510454796),
 (5, 4, 21.75674247468588),
 (6, 4, 12.51912084529485),
 (7, 12, 7.4605341973673704),
 (8, 1, 81.00919920405966),
 (9, 18, 21.56261227624718),
 (10, 15, 377.32135494582764),
 (11, 18, 208.45090235163937),
 (13, 2, 110.6149374875428),
 (14, 19, 291.6134662380013)]

5. Ahora que tienes los datos para cumplir con la solicitud del equipo de marketing, puedes calcular la distancia típica de cada cliente a su concesionario más cercano. Has aprendido que hay dos formas comunes de representar el valor típico de un conjunto de datos, la media y la mediana. Obtendrás ambas utilizando la siguiente consulta:
El resultado es que la distancia promedio es de aproximadamente 147 millas, pero la distancia mediana es de aproximadamente 91 millas.

In [ ]:
%%sql
-- The result is that the average distance is about 147 miles away, but the median distance is about 91 miles.
SELECT
  AVG(distance) AS avg_dist,
  PERCENTILE_DISC(0.5)
    WITHIN GROUP (ORDER BY distance) AS median_dist
FROM
  closest_dealerships;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(146.77826608034164, 91.23958293233487)]



Hay una clara diferencia entre la media y la mediana. Ambas son indicadores importantes de la tendencia central, que representa el valor más típico del conjunto de datos. Pero,
- ¿por qué estas dos medidas típicas para el mismo conjunto de datos son tan diferentes?
- ¿Qué te dice esto sobre los datos?

En el Capítulo 1, Entender y Describir Datos, aprendiste que la media es más sensible a los valores atípicos. Aparentemente, hay algunos clientes cuya distancia al concesionario más cercano es mucho mayor que la de la mayoría de los clientes. Como tal, la media es significativamente mayor que la mediana. Generalmente, es una buena idea calcular tanto la media como la mediana de una variable. Si hay una diferencia significativa en los valores de la media y la mediana, entonces el conjunto de datos puede tener valores atípicos. Necesitas identificar estos valores atípicos para un análisis más detenido.

Al identificar el problema con este conjunto de datos, surge la pregunta de si estos valores atípicos son causados por problemas de calidad de datos o no. Al identificar qué clientes son valores atípicos, revisas su fuente de datos residenciales para confirmar que su dirección registrada es realmente su dirección residencial más actual, no la dirección en la que vivieron hace 10 años. Una vez identificados, utilizarás las técnicas que aprendiste en capítulos anteriores para actualizar la información y volver a ejecutar el análisis, formando así un ciclo de limpieza y mejora de datos. Este es un flujo de trabajo muy común y útil en el campo del análisis de datos. También ayudará a mejorar la calidad de los datos operacionales y a reducir el desperdicio y los errores innecesarios por parte del equipo de operaciones.

Pero, ¿qué pasa si los datos son correctos? ¿Cómo será útil esta información para el negocio? Siempre recuerda que el propósito del análisis de datos es proporcionar información sobre el negocio. Ahora que sabes que algunos clientes viven más lejos de los concesionarios que la mayoría de los clientes, ¿qué decisiones puedes tomar basándote en este conocimiento?
¿Consideras que la existencia de estos clientes indica la necesidad de más concesionarios en su área? Estas observaciones y análisis son exactamente lo que el equipo de gestión espera que hagan los analistas de datos y deben discutirse con el equipo de gestión.

# Usando tipos de datos de Array en PostgreSQL
Mientras que los tipos de datos de PostgreSQL que has explorado hasta ahora te permiten almacenar muchos tipos diferentes de datos, ocasionalmente querrás almacenar una serie de valores en una tabla. Por ejemplo, podrías querer almacenar una lista de los productos que un cliente ha comprado o los números de identificación de empleado asociados con un concesionario específico. Para este escenario, PostgreSQL ofrece el tipo de datos ARRAY, que te permite almacenar una lista de valores.

## Comenzando con Arrays
Los arrays de PostgreSQL te permiten almacenar múltiples valores en un campo en una tabla. Por ejemplo, considera el siguiente primer registro en la tabla de clientes

In [ ]:
%sql SELECT * FROM customers WHERE customer_id = 1;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(1, None, 'Arlena', 'Riveles', None, 'ariveles0@stumbleupon.com', 'F', '98.36.172.246', None, None, None, None, None, None, None, datetime.datetime(2019, 12, 19, 0, 0))]

Cada campo contiene exactamente un valor (el valor NULL sigue siendo un valor); sin embargo, hay algunos atributos que podrían contener múltiples valores con una longitud no especificada. Por ejemplo, supongamos que quisieras tener un campo de productos_comprados. Este podría contener cero o más valores dentro del campo. Imagina que el cliente compró los scooters Lemon y Bat Edición Limitada. Podrías representarlo de la siguiente manera:

Puedes definir un array de varias maneras. Una de las formas de empezar es simplemente creando un array usando el siguiente comando:


In [ ]:
%%sql
SELECT
  ARRAY['Lemon', 'Bat Limited Edition'] AS example_purchased_products;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(['Lemon', 'Bat Limited Edition'],)]

PostgreSQL sabe que los valores Lemon y Bat Limited Edition son del tipo de dato TEXT, por lo que crea un array TEXT para almacenar estos valores.

Aunque puedes crear un array para cualquier tipo de dato, el array está limitado a valores de ese tipo de dato solamente. Así que, no podrías tener un valor entero seguido de un valor de texto o viceversa (esto produciría un error).

También puedes crear arrays usando la función agregada ARRAY_AGG. Esta función agregada creará un array de todos los valores en el grupo. Esto es útil cuando deseas tener una lista consolidada de sub-atributos para cada valor en un atributo principal. Por ejemplo, la siguiente consulta agrega todos los vehículos para cada tipo de producto:

In [45]:
%%sql
SELECT
  product_type, ARRAY_AGG(DISTINCT model) AS models
FROM
  products
GROUP BY 1;

 * postgresql://postgres:***@localhost:5432/sqlda
2 rows affected.


[('automobile', ['Model Chi', 'Model Epsilon', 'Model Gamma', 'Model Sigma']),
 ('scooter', ['Bat', 'Bat Limited Edition', 'Blade', 'Lemon', 'Lemon Limited Edition', 'Lemon Zester'])]

También puedes especificar cómo ordenar los elementos incluyendo una declaración ORDER BY en la función ARRAY_AGG, como en el siguiente ejemplo:


In [46]:
%%sql
SELECT
  product_type,
  ARRAY_AGG(model ORDER BY year) AS models
FROM
  products
GROUP BY 1;

 * postgresql://postgres:***@localhost:5432/sqlda
2 rows affected.


[('automobile', ['Model Chi', 'Model Sigma', 'Model Epsilon', 'Model Gamma', 'Model Chi']),
 ('scooter', ['Lemon', 'Lemon Limited Edition', 'Lemon', 'Blade', 'Bat', 'Bat Limited Edition', 'Lemon Zester'])]

Pero puede haber situaciones en las que quieras revertir esta operación. Esto se puede hacer utilizando la función UNNEST, que crea una fila para cada valor en el array:


In [47]:
 %sql SELECT UNNEST(ARRAY[123, 456, 789]) AS example_ids;

 * postgresql://postgres:***@localhost:5432/sqlda
3 rows affected.


[(123,), (456,), (789,)]

También puedes crear un array dividiendo un valor de cadena utilizando la función STRING_TO_ARRAY. Un escenario común es que, cuando utilizas sistemas de transacciones externas, muchos sistemas actualmente generarán salidas de texto que contienen toda la información en una cadena. Necesitarás dividir la cadena en múltiples partes y analizar cada parte en consecuencia. Aquí hay un ejemplo:

In [48]:
 %sql SELECT STRING_TO_ARRAY('hello there how are you?', ' ');

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(['hello', 'there', 'how', 'are', 'you?'],)]

De manera similar, puedes ejecutar la operación inversa y concatenar un array de cadenas en una sola cadena:


In [49]:
%%sql
SELECT
  ARRAY_TO_STRING(
    ARRAY['Lemon', 'Bat Limited Edition'], ', '
  ) AS example_purchased_products;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[('Lemon, Bat Limited Edition',)]

In [50]:
%sql SELECT gen_random_uuid();

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(UUID('2e0fc004-4839-45e0-91ad-037c322129c5'),)]

Hay otras funciones que te permiten interactuar con arrays. Aquí hay algunos ejemplos de las funcionalidades adicionales de array que proporciona PostgreSQL:

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/8ac068a5-49fe-41bd-8cf3-fc50da73ce83.png?raw=true' width="600" />
<figcaption>
Figure 7.9: Examples of additional array functionality</figcaption></center>
</figure>

# Ejercicio 7.03: Analizando Secuencias Usando Arreglos
En este ejercicio, utilizarás arreglos para analizar secuencias. ZoomZoom envía correos electrónicos a los clientes en series. Por ejemplo, antes de la temporada de festividades de diciembre, enviarán un correo electrónico proporcionando un catálogo de productos de todo lo que venden. Durante la temporada, enviarán actualizaciones sobre qué producto se está vendiendo bien y qué descuentos se ofrecen. Después de la temporada, enviarán correos electrónicos de agradecimiento y ofrecerán más productos y descuentos. El equipo de marketing quiere que identifiques las tres secuencias de correo electrónico más comunes. Les ayudarás a entender mejor cuán diferentes son estas secuencias al observar si estas secuencias son superconjuntos una de la otra:

1. Primero, crea una tabla que represente la secuencia de correos electrónicos para cada cliente:

In [68]:
# %sql DROP TABLE customer_email_sequences;
%%sql
CREATE TEMP TABLE customer_email_sequences AS (
  SELECT
    customer_id,
    ARRAY_AGG(
      email_subject ORDER BY sent_date
    ) AS email_sequence
FROM emails
GROUP BY 1
);

 * postgresql://postgres:***@localhost:5432/sqlda
47437 rows affected.


[]

In [70]:
%%sql
SELECT
    customer_id,
    ARRAY_AGG(
      email_subject ORDER BY sent_date
    ) AS email_sequence
FROM emails
GROUP BY 1 LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, ['Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (2, ['Zoom Zoom Black Friday Sale', 'An Electric Car for a New Age', "Tis' the Season for Savings", 'Like a Bat out of Heaven', "25% off all EVs. It's a C ... (16 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (3, ['Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (4, ['Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (5, ['A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Blade', 'Zoom Zoom Black Friday Sale', 'An Electric Car for a New Age', "Tis' the Sea ... (88 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (6, ['Shocking Holiday Savings On Electric Scooters', 'A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Blade', 'Zoom Zoom Black Friday Sale ... (137 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (7, ['The 2013 Lemon Scooter is Here', 'Shocking Holiday Savings On Electric Scooters', 'A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Bl ... (171 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (8, ['A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Blade', 'Zoom Zoom Black Friday Sale', 'An Electric Car for a New Age', "Tis' the Sea ... (88 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (9, ['Like a Bat out of Heaven', "25% off all EVs. It's a Christmas Miracle!", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (10, ['The 2013 Lemon Scooter is Here', 'Shocking Holiday Savings On Electric Scooters', 'A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Bl ... (171 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs'])]

2. Luego, identifica las tres secuencias de correo electrónico más comunes. Dado que ya tienes las secuencias de correo electrónico, puedes hacer esto utilizando ORDER BY con LIMIT 3. Como la cláusula ORDER BY se basa en la ocurrencia de secuencias de correo electrónico, la instrucción SELECT mostrará las secuencias con las más frecuentes primero. Luego, con la cláusula LIMIT 3, la instrucción devolverá solo las top 3 secuencias:

In [75]:
%%sql
SELECT
COUNT(1) AS occurrences,
    email_sequence

  FROM
    customer_email_sequences
  GROUP BY email_sequence
  ORDER BY occurrences
  DESC LIMIT 3


 * postgresql://postgres:***@localhost:5432/sqlda
3 rows affected.


[(10323, ['The 2013 Lemon Scooter is Here', 'Shocking Holiday Savings On Electric Scooters', 'A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Bl ... (171 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (5340, ['Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs']),
 (4631, ['Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs'])]

In [ ]:
%%sql
CREATE TEMP TABLE top_email_sequences AS (
  SELECT
    email_sequence,
    COUNT(1) AS occurrences
  FROM
    customer_email_sequences
  GROUP BY 1
  ORDER BY 2
  DESC LIMIT 3
 );

3. Finalmente, querrás verificar cuál de estos arrays es un superconjunto de los otros arrays. Es posible que algunos clientes se hayan unido más tarde que otros, por lo que solo recibieron una parte de la secuencia de correo electrónico. Necesitas identificar estas sub-secuencias como parte de la secuencia completa de correo electrónico.

- El único problema es que los campos de secuencia de correo electrónico son muy largos y no son intuitivos para leer con ojos humanos. Para ayudar con esto, es útil asignar a tus filas un ID numérico para identificación:

https://es.wikipedia.org/wiki/Teor%C3%ADa_de_conjuntos#/media/Archivo:Venn_A_intersect_B.svg

In [65]:
%%sql
ALTER TABLE
  top_email_sequences
ADD COLUMN
  id SERIAL PRIMARY KEY;

 * postgresql://postgres:***@localhost:5432/sqlda
(psycopg2.errors.DuplicateColumn) column "id" of relation "top_email_sequences" already exists

[SQL: ALTER TABLE
  top_email_sequences
ADD COLUMN
  id SERIAL PRIMARY KEY;]
(Background on this error at: https://sqlalche.me/e/20/f405)


In [77]:
%%sql SELECT * FROM top_email_sequences

 * postgresql://postgres:***@localhost:5432/sqlda
3 rows affected.


[(['The 2013 Lemon Scooter is Here', 'Shocking Holiday Savings On Electric Scooters', 'A Brand New Scooter...and Car', 'We cut you a deal: 20%% off a Bl ... (171 characters truncated) ... !", 'We Really Outdid Ourselves this Year', 'Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs'], 10323, 1),
 (['Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs'], 5340, 2),
 (['Black Friday. Green Cars.', 'Save the Planet with some Holiday Savings.', 'A New Year, And Some New EVs'], 4631, 3)]

4. A continuación, puedes hacer una unión cruzada de la tabla consigo misma y usar el operador @> para verificar si un array que contiene una secuencia de correos electrónicos contiene otra secuencia de correos electrónicos:

In [ ]:
%%sql
SELECT
  super_email_seq.id AS superset_id,
  sub_email_seq.id AS subset_id
FROM
  top_email_sequences AS super_email_seq
  CROSS JOIN
  top_email_sequences AS sub_email_seq
WHERE
  super_email_seq.email_sequence @> sub_email_seq.email_sequence
AND
  super_email_seq.id != sub_email_seq.id;

 * postgresql://postgres:***@localhost:5432/sqlda
3 rows affected.


[(1, 2), (1, 3), (3, 2)]

De esto, puedes deducir que la secuencia de correos electrónicos más común contiene la segunda y tercera secuencias de correo electrónicos más comunes, mientras que la tercera secuencia de correos electrónicos más común es un superconjunto de la segunda secuencia más común. Este tipo de análisis generalmente es útil cuando se observa qué puntos de contacto con el cliente podrían llevar a alguien a realizar una compra o no. Por ejemplo, algunos clientes que se unieron tarde pueden no haber recibido el primer correo electrónico, el catálogo de productos de la temporada de fiestas. Pero si un porcentaje similar de estos clientes y los clientes que recibieron el primer correo electrónico realizaron una compra después de recibir el correo electrónico de descuento de la temporada de fiestas, podrías sospechar razonablemente que el correo electrónico de descuento de la temporada de fiestas es la principal razón para la compra, no el catálogo de productos.

- Esto también es conocido como **modelado de atribución**.

#Uso de tipos de datos JSON en PostgreSQL
Aunque los arrays pueden ser útiles para almacenar una lista de valores en un solo campo, a veces
tus estructuras de datos pueden ser complejas.
- Podrías querer almacenar múltiples valores de diferentes tipos en un solo campo, y
- podrías querer que los datos estén etiquetados con labels en lugar de almacenados secuencialmente.

Estos son problemas comunes con datos a nivel de registro, así como datos alternativos.

- Por ejemplo, una base de datos de pacientes de atención médica puede contener un campo llamado receta, que contiene todas las recetas de un paciente. Algunos pacientes pueden no tener ninguna receta, por lo que este campo puede estar vacío. Otros pacientes pueden tener múltiples recetas, y la receta de cada paciente puede ser diferente de las demás. Un paciente puede tener un medicamento para la hipertensión de 10 mg por día. Otro puede tener un medicamento para el insomnio de dos pastillas por noche. Otro paciente más puede tener ambos.

Es muy difícil almacenar estos datos en un formato predefinido, por lo que generalmente se almacenan como pares clave-valor utilizando el formato JSON.

**La Notación de Objetos de JavaScript (JSON)** es un formato de texto estándar abierto para almacenar datos de variada complejidad. Puede ser utilizado para representar casi cualquier cosa, como la información de pacientes de atención médica que viste anteriormente. Esto es diferente del tipo de datos ARRAY, que puede almacenar múltiples valores. Los valores deben ser del mismo tipo. Una tabla de base de datos tiene nombres de columnas, mientras que los datos JSON tienen claves. Puedes utilizar JSON para representar un registro de tu tabla de clientes fácilmente almacenando nombres de columnas como claves y valores de fila como valores. La función row_to_json transforma filas a JSON:

In [80]:
%sql SELECT * FROM customers c limit 1;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(716, None, 'Jarred', 'Bester', None, 'jbesterjv@nih.gov', 'M', '216.51.110.28', None, None, None, None, None, None, None, datetime.datetime(2018, 9, 19, 0, 0))]

In [81]:
%sql SELECT row_to_json(c) FROM customers c limit 1;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[({'customer_id': 716, 'title': None, 'first_name': 'Jarred', 'last_name': 'Bester', 'suffix': None, 'email': 'jbesterjv@nih.gov', 'gender': 'M', 'ip_ad ... (37 characters truncated) ... e, 'street_address': None, 'city': None, 'state': None, 'postal_code': None, 'latitude': None, 'longitude': None, 'date_added': '2018-09-19T00:00:00'},)]

Esto es un poco difícil de leer, pero puedes agregar el indicador pretty_bool para generar una versión legible. En la siguiente consulta, el segundo parámetro de la función row_to_json es el indicador pretty_bool y está configurado como TRUE:

In [82]:
%sql SELECT jsonb_pretty(row_to_json(c)::jsonb) FROM customers c LIMIT 1;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[('{\n    "city": null,\n    "email": "jbesterjv@nih.gov",\n    "phone": null,\n    "state": null,\n    "title": null,\n    "gender": "M",\n    "suffix" ... (122 characters truncated) ... ",\n    "first_name": "Jarred",\n    "ip_address": "216.51.110.28",\n    "customer_id": 716,\n    "postal_code": null,\n    "street_address": null\n}',)]